### PR TITLE
Add yield_for_timeout Function

### PR DESCRIPTION
**Pull request by adkinsjd**
This PR adds a

```int yield_for_timeout(bool* cond, uint32_t ms)```

function to the timer library. It acts like yield_for, however it will return early if the condition is not met before the timeout. The return code is TOCK_SUCCESS if it returns due to the condition being met, and TOCK_FAIL if it returns due to a timeout. I believe this will be generally useful for waiting on external events. There is also a test case which I used to test the function.

Some considerations:
 - There was some discussion about this going in tock.h instead of timer.h, however putting it in 
 timer.h requires adding no more dependencies to tock.h
 - I briefly thought about the addition of another tock error code 'TOCK_TIMEOUT', but I was not sure if it would be generally useful.

### CODE REVIEW

**Comment by ppannuto** [comment]
Little things first:

 - I think `yield_for_with_timeout` might be a better name, when I first read through the example, I got confused for a second because I thought it was yielding for a timeout, which didn't make sense
 - I think `TOCK_ETIMEDOUT` would be a generally useful error code, but FAIL for now is fine

Possibly bigger problem:

I'm not sure this can be done safely in user space, I think timeout may need to be a parameter to the yield syscall. I was trying to think through what happens if the timeout and the event happen around roughly the same time. This gets into the kernel implementation a bit. Specifically, consider this sequence:

1. The timer ISR is running in the kernel
2. The interrupt for the event you actually want fires, it's pended while the time ISR finishes
3. The timer ISR schedules the timer callback to run in the process
4. The event ISR runs, and schedules the event callback to run in the process
5. The kernel allows the process to run, popping the first callback, which is the timer callback which runs
6. When the timer callback finishes, [this yield](https://github.com/helena-project/tock/pull/473/files#diff-fa822b1ced5e2f4bc871b880e77f14edR261) will resume execution here - note the problem, the event callback won't run until the _next_ time the process yields
7. `cond` will still (wrongly) be false, and the process will continue believing the timeout tripped when the event actually completed successfully

**Comment by adkinsjd** [comment]
If the timer ISR is running in the kernel then the event technically occurred after the timeout correct?

So while the event may occur before the function returns a failure, the function will not return a failure if the event occurs before the timeout. Is my thinking correct? Is this a sufficient guarantee?

**Comment by alevy** [comment]
I think I agree with @adkinsjd on this. There is never a way to guarantee that an event doesn't occur after a timeout elapses (you could _maybe_ imagine doing this with sleep walking I guess), so it's just a question of how likely it is to have happened despite the timeout, which yeah, is higher if this is implemented in user space, but the semantics of the call don't change.

**Comment by adkinsjd** [comment]
@alevy do you agree with the name change and the addition of a new error code?

**Comment by alevy** [comment]
Name change yes, additional error code.... less so. I wouldn't not approve but I'm not as stokes to add new error codes.

File: userland/examples/tests/yield_for_with_timeout_test/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Test yield_for_timeout
```
Comment (ppannuto) [inline userland/examples/tests/yield_for_with_timeout_test/README.md]: name fixes in this file :)

**Comment by alevy** [comment]
@ppannuto you merged before letting travis finish, and there is a formatting error...

**Comment by alevy** [comment]
Actually, the problem was probably in #472, which was also merged before the CI finished.

**Comment by ppannuto** [comment]
My mistake, I could've sworn that had finished, but apparently there are
two strikes of evidence to the contrary, sorry

**Pull request merged**